src/heretic/tagger.py

```diff
@@ -0,0 +1,60 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# Copyright (C) 2025  Philipp Emanuel Weidmann <pew@worldwidemann.com>
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any, Dict
+
+from heretic.schemas import ContextMetadata, ResponseMetadata
+
+if TYPE_CHECKING:
+    from .config import Settings
+    from .model import Model
+
+
+class Tagger(ABC):
+    """
+    Abstract base class for tagger plugins.
+    Taggers take in LLM responses + optional metadata and output a map of attributes, e.g:
+    "I'm sorry, but I can't assist with that request." -> {"helpfulness": 0.1, "friendliness": 0.2}
+    """
+
+    def __init__(
+        self,
+        settings: "Settings",
+        model: "Model",
+        context_metadata: ContextMetadata,
+    ):
+        self.settings = settings
+        self.model = model
+        self.context_metadata = context_metadata
+
+    @abstractmethod
+    def tag_batch(
+        self, responses: list[str], metadata: list[ResponseMetadata]
```

Comment:
It's probably cleaner to have a single object `Response` that contains all data, including the text.